llvm/lib/Transforms/Utils/SYCLModuleSplit.cpp

```diff
@@ -0,0 +1,935 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// See comments in the header.
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/SYCLModuleSplit.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Bitcode/BitcodeWriterPass.h"
+#include "llvm/Demangle/Demangle.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/IR/PassManagerImpl.h"
+#include "llvm/IRPrinter/IRPrintingPasses.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/LineIterator.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Transforms/IPO.h"
+#include "llvm/Transforms/IPO/GlobalDCE.h"
+#include "llvm/Transforms/IPO/Internalize.h"
+#include "llvm/Transforms/IPO/StripDeadPrototypes.h"
+#include "llvm/Transforms/IPO/StripSymbols.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include "llvm/Transforms/Utils/SYCLUtils.h"
+
+#include <algorithm>
+#include <map>
+#include <utility>
+#include <variant>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "sycl_module_split"
+
+namespace {
+// Identifying name for global scope
+constexpr char GLOBAL_SCOPE_NAME[] = "<GLOBAL>";
+constexpr char SYCL_SCOPE_NAME[] = "<SYCL>";
+
+EntryPointsGroupScope selectDeviceCodeGroupScope(const Module &M,
+                                                 IRSplitMode Mode,
+                                                 bool AutoSplitIsGlobalScope) {
+  switch (Mode) {
+  case SPLIT_PER_TU:
+    return Scope_PerModule;
+
+  case SPLIT_PER_KERNEL:
+    return Scope_PerKernel;
+
+  case SPLIT_AUTO: {
+    if (AutoSplitIsGlobalScope)
+      return Scope_Global;
+
+    // At the moment, we assume that per-source split is the best way of
+    // splitting device code and can always be used except for cases handled
+    // above.
+    return Scope_PerModule;
+  }
+
+  case SPLIT_NONE:
+    return Scope_Global;
+  }
+
+  llvm_unreachable("unsupported split mode");
+}
+
+// Return true if the function is a SPIRV or SYCL builtin, e.g.
+// _Z28__spirv_GlobalInvocationId_xv
+bool isSpirvSyclBuiltin(StringRef FName) {
+  if (!FName.consume_front("_Z"))
+    return false;
+  // now skip the digits
+  FName = FName.drop_while([](char C) { return std::isdigit(C); });
+
+  return FName.starts_with("__spirv_") || FName.starts_with("__sycl_");
+}
+
+// Return true if the function name starts with "__builtin_"
```

Comment:
Do we really have function like this in use? Would it be safe to assume that _any_ function whose name starts with `__` should not be ever considered as an entry point? Those names are reserved for the implementation according to C++ spec and they can't be used by SYCL users.